From a075be8d6f3bacea9a7ccc36f80bac2e3276c5e2 Mon Sep 17 00:00:00 2001
From: Kevin Hao <kexin.hao@windriver.com>
Date: Tue, 20 Aug 2019 11:22:13 +0800
Subject: [PATCH 309/767] octeontx2-af: Fix the using of variable length arrays

The Variable Length Arrays (VLAs) is not allowed to be used in the
kernel codes. And a warning will be ejected after commit 8289f913fe126
("kbuild: add -Wvla flag unconditionally").

Signed-off-by: Kevin Hao <kexin.hao@windriver.com>
---
 drivers/net/ethernet/marvell/octeontx2/af/rvu.c | 14 ++++++++++++--
 1 file changed, 12 insertions(+), 2 deletions(-)

diff --git a/drivers/net/ethernet/marvell/octeontx2/af/rvu.c b/drivers/net/ethernet/marvell/octeontx2/af/rvu.c
index 93823c21813d..a6855d86993f 100644
--- a/drivers/net/ethernet/marvell/octeontx2/af/rvu.c
+++ b/drivers/net/ethernet/marvell/octeontx2/af/rvu.c
@@ -2033,16 +2033,24 @@ static void rvu_npa_lf_mapped_nix_lf_teardown(struct rvu *rvu, u16 pcifunc)
 
 static void rvu_npa_lf_mapped_sso_lf_teardown(struct rvu *rvu, u16 pcifunc)
 {
-	u16 pcifunc_arr[rvu->hw->total_pfs + rvu->hw->total_vfs];
+	u16 *pcifunc_arr;
 	u16 sso_pcifunc, match_cnt = 0;
 	struct rvu_block *sso_block;
 	struct rsrc_detach detach;
 	int blkaddr, lf;
 	u64 regval;
+	size_t len;
+
+	len = sizeof(*pcifunc_arr) * (rvu->hw->total_pfs + rvu->hw->total_vfs);
+	pcifunc_arr = kmalloc(len, GFP_KERNEL);
+	if (!pcifunc_arr)
+		return;
 
 	blkaddr = rvu_get_blkaddr(rvu, BLKTYPE_SSO, 0);
-	if (blkaddr < 0)
+	if (blkaddr < 0) {
+		kfree(pcifunc_arr);
 		return;
+	}
 
 	sso_block = &rvu->hw->block[blkaddr];
 	for (lf = 0; lf < sso_block->lf.max; lf++) {
@@ -2067,6 +2075,8 @@ static void rvu_npa_lf_mapped_sso_lf_teardown(struct rvu *rvu, u16 pcifunc)
 
 	for (sso_pcifunc = 0; sso_pcifunc < match_cnt; sso_pcifunc++)
 		rvu_detach_rsrcs(rvu, &detach, pcifunc_arr[sso_pcifunc]);
+
+	kfree(pcifunc_arr);
 }
 
 static void rvu_blklf_teardown(struct rvu *rvu, u16 pcifunc, u8 blkaddr)
-- 
2.31.1

